cmd/gateway/commands.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/config"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/manager"
+)
+
+const (
+	domain                = "k8s-gateway.nginx.org"
+	gatewayClassFlag      = "gatewayclass"
+	gatewayClassNameUsage = `The name of the GatewayClass resource. ` +
+		`Every NGINX Gateway must have a unique corresponding GatewayClass resource.`
+	gatewayCtrlNameFlag     = "gateway-ctlr-name"
+	gatewayCtrlNameUsageFmt = `The name of the Gateway controller. ` +
+		`The controller name must be of the form: DOMAIN/PATH. The controller's domain is '%s'`
+)
+
+var (
+	// Backing values for common cli flags shared among all subcommands
+	// The values are managed by the Root command.
+	gatewayCtlrName = stringValidatingValue{
+		validator: validateGatewayControllerName,
+	}
+
+	gatewayClassName = stringValidatingValue{
+		validator: validateResourceName,
+	}
+)
+
+// stringValidatingValue is a string flag value with custom validation logic.
+// stringValidatingValue implements the pflag.Value interface.
+type stringValidatingValue struct {
+	validator func(v string) error
+	value     string
+}
+
+func (v *stringValidatingValue) String() string {
+	return v.value
+}
+
+func (v *stringValidatingValue) Set(param string) error {
+	if err := v.validator(param); err != nil {
+		return err
+	}
+	v.value = param
+	return nil
+}
+
+func (v *stringValidatingValue) Type() string {
+	return "string"
+}
+
+func createRootCommand() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use: "gateway",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cmd.Help()
+		},
+	}
+
+	rootCmd.PersistentFlags().Var(
+		&gatewayCtlrName,
+		gatewayCtrlNameFlag,
+		fmt.Sprintf(gatewayCtrlNameUsageFmt, domain),
+	)
+	utilruntime.Must(rootCmd.MarkPersistentFlagRequired(gatewayCtrlNameFlag))
+
+	rootCmd.PersistentFlags().Var(
+		&gatewayClassName,
+		gatewayClassFlag,
+		gatewayClassNameUsage,
+	)
+	utilruntime.Must(rootCmd.MarkPersistentFlagRequired(gatewayClassFlag))
+
+	return rootCmd
+}
+
+func createControlPlaneCommand() *cobra.Command {
+	return &cobra.Command{
+		Use:   "control-plane",
```

Comment:
What about `controller`?

Reply:
> What about controller?

I started renaming but then realized we have the `controller` package. I know the command is user facing, but the source code will also match it, and we already have a concept of a controller in the code.

At the same time, I'm not a big fan of `control-plane` but can't think of a better name.

Perhaps you have more alternatives in mind? 
@kate-osborn  ?

Reply:
perhaps `configurer` as in "Configure NGINX according with resources creating in the cluster" ?

Reply:
What about `nginx-gateway` and `nginx-gateway-provisioner`?

Reply:
it's gonna be `./gateway nginx-gateway ...` and `./gateway nginx-gateway-provisioner ...` . 
at the same time, I like it amongst the other options considered

Reply:
Or `default-mode` and `provisioner-mode`